### ISSUE TITLE
Plans for training custom model on ai platform and preemptible

### ISSUE THREAD

**Issue by dkajtoch** [google-research/text-to-text-transfer-transformer#549]
Hi, I am planning to pre-train t5 base model on some custom dataset. I am looking for information on how to do it properly and understand how much resources do I need (TPU cores). Here are couple of questions I have:

1. In the paper you used batch size of 128 examples and you also mention somewhere that you use TPU v3-1024 so 1024 cores. The base model is not that big so I wonder what would be the benefit of running it on so many cores with effective batch size of just 128 (total I guess)? What would be the good tpu size for training base model?
2. Did you truly trainer with batch size 128? Can I scale it let's say 8192 and use more data parallelism (what about i/o stuff then)?
3. My dataset is just text lines split into multiple parts. I was planing to write `TextLineTask`but found it may slow down the training: [https://github.com/google-research/text-to-text-transfer-transformer/issues/282#issuecomment-651382177](https://github.com/google-research/text-to-text-transfer-transformer/issues/282#issuecomment-651382177) Should I switch to tfds for base model?
4. So far I run models on GPUs and used ai platform to schedule training. I wonder if it is possible to use you library and schedule custom training on ai platform with `gcloud` util?
5. Generally, preemptible machines are cheaper so it would be nice to run the training on them. There is of course a problem with machine going down and recovering from failure. Is it possible in your library to recover training automatically from latest checkpoint?

 Thanks for clarifying that!

**Comment by craffel** [comment]
> you use TPU v3-1024

That was for training T5-11B. We used much smaller slices for the smaller models. v3-32 is plenty big for Base.

> Did you truly trainer with batch size 128? Can I scale it let's say 8192 and use more data parallelism (what about i/o stuff then)?

We used a batch size of 65536 input tokens = 128 sequences/batch for most of the ablation experiments in the paper. For the T5 models at the end of the paper, we used a batch size of 2^20 tokens = 2048 sequences/batch. That batch size is plenty big to take advantage of plenty of data parallelism.

> My dataset is just text lines split into multiple parts. I was planing to write TextLineTaskbut found it may slow down the training: #282 (comment) Should I switch to tfds for base model?

You should cache the Task.

> So far I run models on GPUs and used ai platform to schedule training. I wonder if it is possible to use you library and schedule custom training on ai platform with gcloud util?

I don't know, I've never used AI platform.

> Generally, preemptible machines are cheaper so it would be nice to run the training on them. There is of course a problem with machine going down and recovering from failure. Is it possible in your library to recover training automatically from latest checkpoint?

Yes, this is what it does by default.

**Comment by dkajtoch** [comment]
Thank you very much for the response! One more thing, is dataset state part of the checkpoint? If the machines dies and I would like to restart training from latest checkpoint will it "rewind the dataset" or start from the beginning?

**Comment by adarob** [comment]
If you use a seed in the mesh_train_dataset_fn and set skip_seen_data to True, it will continue the dataset where you left off

https://github.com/tensorflow/mesh/blob/f6aceac90dbe2602b1478270c43690308a599079/mesh_tensorflow/transformer/utils.py#L1610

**Comment by marcospiau** [comment]
Hi @adarob. Regarding deterministic training, do you know what happens when I run multiple passes over the training dataset? Are the training examples sampled again before every epoch, or every epoch uses exactly the same examples? I already have a cached training task. Thanks in advance!
